collector/iscsi_linux.go

```diff
@@ -0,0 +1,402 @@
+// Copyright 2019 The Prometheus Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package collector
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/log"
+	"github.com/prometheus/procfs/iscsi"
+)
+
+const (
+	lioFileioSubsystem = "lio_fileio"
+	lioIblockSubsystem = "lio_iblock"
+	lioRbdSubsystem    = "lio_rbd"
+	lioRdmcpSubsystem  = "lio_rdmcp"
+)
+
+// An lioCollector is a Collector which gathers iscsi RBD
+// iops (iscsi commands) , Read in byte and Write in byte.
+// ( original reading sysfs is in MB )
+type lioCollector struct {
+	fs      iscsi.FS
+	metrics *lioMetric
+}
+
+type lioMetric struct {
+	lioFileIops  *prometheus.Desc
+	lioFileRead  *prometheus.Desc
+	lioFileWrite *prometheus.Desc
+
+	lioBlockIops  *prometheus.Desc
+	lioBlockRead  *prometheus.Desc
+	lioBlockWrite *prometheus.Desc
+
+	lioRbdIops  *prometheus.Desc
+	lioRbdRead  *prometheus.Desc
+	lioRbdWrite *prometheus.Desc
+
+	lioRdmcpIops  *prometheus.Desc
+	lioRdmcpRead  *prometheus.Desc
+	lioRdmcpWrite *prometheus.Desc
+}
+
+type graphLabel struct {
+	iqn   string
+	tpgt  string
+	lun   string
+	store string
+	pool  string
+	image string
+}
+
+func init() {
+	registerCollector("iscsi", defaultEnabled, NewLioCollector)
+}
+
+// NewLioCollector returns a new Collector with iscsi statistics.
+func NewLioCollector() (Collector, error) {
+	fs, err := iscsi.NewFS(*sysPath, *configfsPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open sysfs / configfs: %v", err)
+	}
+
+	metrics, _ := newLioMetric()
+
+	return &lioCollector{
+		fs:      fs,
+		metrics: metrics}, nil
+}
+
+// Update implement the lioCollector.
+func (c *lioCollector) Update(ch chan<- prometheus.Metric) error {
+
+	stats, err := c.fs.ISCSIStats()
+	log.Debugf("lio: Update lioCollector")
+	if err != nil {
+		return fmt.Errorf("lio: failed to update iscsi stat : %v", err)
+	}
+	for _, s := range stats {
+		c.updateStat(ch, s)
+	}
+	return nil
+}
+
+//newLioMetric create the LIO metric data structure to return for node_exporter
+func newLioMetric() (*lioMetric, error) {
```

Comment:
Looks like `error` can be removed from the return type, since it will always be nil.

Reply:
I think I was just following other collector as my example. Most of the New metric function doesn't really return real error but nil. You are correct, I can either find an example to return a real error or just remove it for now.